### PR TITLE
Add API to enable comparision of corefx's OpenSSL with a 3rd party P/Invoke

### LINKED ISSUE #32718: API to detect the bound version of OpenSSL
Since SafeEvpPKeyHandle exists to allow interop with OpenSSL, it would be good for callers to be able to determine if they are using the same library/version of OpenSSL that we are.

```C#
    public sealed partial class SafeEvpPKeyHandle
    {
        // The result of SSLeay() (OpenSSL 1.0.2) or OpenSSL_version() (OpenSSL 1.1.x)
        public static int OpenSslVersion { get; }
    }
```

Discussion points:
* The native API is Unix C-long, so 32-bit on 32-bit, 64-bit on 64-bit.
* So, `int`, `uint`, `long`, or `ulong`?
* The version number contains detail we can't expose via System.Version
* Callers are only expected to do `==`

### PR DESCRIPTION
**Pull request by bartonjs**
This adds SafeEvpPKeyHandle.OpenSslVersion.

3rd party code which is doing a P/Invoke into OpenSSL (direct or via their own shim) and using RSAOpenSsl..ctor(SafeEvpPKeyHandle) (or DSAOpenSsl or ECDsaOpenSsl) is encouraged to invoke the appropriate version function for their library (`SSLeay()` or `OpenSSL_version_number()`) and compare that against the value of `SafeEvpPKeyHandle.OpenSslVersion`.

As a structural change, this also changes the shim method from a 32-bit to a 64-bit answer, to ensure that we don't truncate in the future if 64-bit OpenSSL changes versioning schemes to exceed a 32-bit value (the native API returns a \*NIX C long).

Fixes #32718.

### CODE REVIEW

File: src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSslVersion.cs [hunk 1]
```diff
@@ -12,16 +12,28 @@ internal static partial class OpenSsl
...
+                    // e.g. 1.0.2a final is 0x1000201F
+                    //
+                    // Currently they don't exceed 29-bit values, but we use long here to account
+                    // for the expanded range on their 64-bit C-long return value.
+                    long versionNumber = OpenSslVersionNumber();
+                    int major = (int)((versionNumber >> 28) & 0xF);
+                    int minor = (int)((versionNumber >> 20) & 0xFF);
+                    int fix = (int)((versionNumber >> 12) & 0xFF);
+
+                    s_opensslVersion = new Version(major, minor, fix);
```
Comment (stephentoub) [inline src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSslVersion.cs]: Where is this used?  I'm just surprised to see us parsing it like this, given that we explicitly opted to expose a long due to it being opaque data.  Is this just for test code?  And if so, should this OpenSslVersion property be moved into a test assembly somewhere?

Comment (bartonjs) [reply src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSslVersion.cs]: >  Is this just for test code?

Yeah.

>  And if so, should this OpenSslVersion property be moved into a test assembly somewhere?

What a remarkably reasonable question that means I have more work to do.  Okie dokie :smile:.

**Comment by bartonjs** [comment]
@dotnet-bot Test Linux arm64 Release Build please (System.Net.Internals.SocketExceptionFactory+ExtendedSocketException : Resource temporarily unavailable during NameResolution tests)

**Comment by stephentoub** [comment]
@dotnet-bot Test Linux arm64 Release Build please
@dotnet-bot test Packaging All Configurations x64 Debug Build please

**Pull request merged**